Revised docs for seq, unfold, from_loop in LazyList

mars0i opened the pull request:
The start of the `seq` docs are now molded on start of the docs for `unfold` and `from_loop`, which I think was clearer.  Also added a note about how to create an infinite list.

`unfold`, `from_loop`: Added explanation of the signature of `next` and what it's for.  Added fibonacci example (which @gasche 

I also reordered the signatures so that the documentation would appear in an order that helps understanding a little bit.  `seq` is first because it's the simpler version, and `unfold` and `from_loop` are now adjacent since they do almost the same thing.  (`from_loop` is just a wrapper around `unfold`.)

Similar changes might be relevant for other modules such as `Enum`, but that can wait for another PR.

gasche commented on src/batLazyList.mli:
```diff
@@ -105,22 +105,54 @@ val from_while: (unit -> 'a option) -> 'a t
+   result, etc. The list ends whenever [next] returns [None].  The function 
+   [next] should return a pair [option] whose first element will be the
+   current value of the sequence; the second element will be passed 
+   (lazily) to [next] in order to compute the following element.  One use 
```
One example use

gasche commented on src/batLazyList.mli:
```diff
@@ -105,22 +105,54 @@ val from_while: (unit -> 'a option) -> 'a t
+   element will be passed (lazily) to [next] in order to compute the following
+   element.  One use of [unfold] is to make each element of the resulting sequence 
+   to depend on the previous two elements, as in this definition of a Fibonacci 
+   sequence:
```
This is very redundant with the documentation of `unfold` above. I think that you could stop after

> The list ends whenever the function raises {!LazyList.No_more_elements}.

and then just refer to the fact that otherwise it behaves as `unfold`.

mars0i commented:
In the `unfold` doc I added "example" before "use", but worded it a little bit differently.

For `from_loop`, I deleted most of the redundant material, as suggested, but added a note referencing `unfold` with a suggestion about what's different.  

I find `unfold` and `from_loop` initially difficult to understand.  I know that the `option`/no `option`' difference should be obvious from the signature, and is what one might expect given that `unfold` doesn't expect an exception while `from_loop` does.  However, for someone like me, at this stage, I think it's helpful to have the absence of `option` in `from_loop` made explicit because of the way in which these functions are already confusing without an additional twist.  That's the rationale for the last comment in the new `from_loop` doc.

Happy to go along with whatever seems best to others, though.

gasche commented:
Status report: I like the new state and would like to merge it, but the commit trail still has a terrible structure, so I need to do the work of cleaning it up (I'll publish a step-by-step log) before, and I've been drowned under other things to do. Ping me again if it's not done next week.

mars0i commented:
On this PR I didn't perform the `cherry_pick` operation you suggested on the other two PRs.  Happy to do that if you want.

mars0i commented:
Github now allows "squashing" multiple commits into a single commit.  It looks like this has to done in the main repo, or enabled there, at least  https://github.com/blog/2141-squash-your-commits

gasche commented:
Thanks for this contribution!
